kaiachain/kaia
# test: Add Shanghai tests by `ethereum/execution-spec-tests`

## Proposed changes

This PR adds tests by `ethereum/execution-spec-tests`. Currently, this supports testing only Shanghai EIPs against each forks (only Cancun and Prague). I'm going to add tests for other forks in other PRs.

Simulating Ethereum behavior is needed because of the difference of spec between Kaia and Ethereum to use `ethereum/execution-spec-tests` as is. This simulate following behavior:

- Gas price - Simulate Ethereum gas prices and set the values ​​directly on the txContext
- Intrinsic gas - Ethereum’s intrinsic gas after Istanbul is the same as Kaia’s Prague
- Op code gas - Change constant gas for specific op code.
- Mining reward - Simualte Ethereum mining reward and set the value directly to the state db.
- State root - Fetch all data from the state db and set the data to a new state db as LegacyAccount

Main code modifications:

- Add SetLegacyAccountForTest to StateDB
- Add relaxPrecompileRangeForTest
- Add ChangeGasCostForTest

Remarks:

- TestExecutionSpecState should be tested separately because of changing global variable (relaxPrecompileRangeForTest)

## Types of changes


## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/kaiachain/kaia/blob/main/CONTRIBUTING.md) doc
- [x] I have read the [CLA](https://gist.github.com/kaiachain-dev/bbf65cc330275c057463c4c94ce787a6) and signed by comment `I have read the CLA Document and I hereby sign the CLA` in first time contribute
